[14.0] Add shopfloor_reception

mmequignon opened the pull request:
Shopfloor implementation of the reception scenario.
Allows to receive products and create the proper packs for each logistic unit.

Depends on this PR https://github.com/OCA/stock-logistics-workflow/pull/1116

ref: rau-49 + rau-57

JuMiSanAr committed:
BSRAU-57: shopfloor_reception_mobile frontend

mmequignon committed:
Add shopfloor_reception

TDu commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,877 @@
+        filtered_lines = move_lines.filtered(
+            lambda l: l.move_id.picking_id.state == "assigned"
+        )
+        filtered_lines = filtered_lines.filtered(
```
Grouping all filtering conditions in one call to the `filtered` function is probably better for performance.

TDu commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,877 @@
+            - set_lot: a single picking has been found for this packaging
+            - select_document: A single or no pickings has been found for this packaging
+        """
+        move_lines = self._filter_move_lines(self._move_line_by_product(product))
```
This will return all existing move lines for a product without any other restriction (that is a lot of lines)
It should filter on the picking types from the shopfloor menu and their state.

TDu commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,877 @@
+        # If no available picking with the right state has been found,
+        # return an error
+        return self._response_for_select_document(
+            message={
```
Messages should be added to `.../actions/message.py`

TDu commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,877 @@
+            - set_lot: a single picking has been found for this packaging
+            - select_document: A single or no pickings has been found for this packaging
+        """
+        move_lines = self._move_line_by_packaging(packaging)
```
Same problem than before with the products.

sebalix commented on shopfloor_reception/__manifest__.py:
```diff
@@ -0,0 +1,20 @@
+    "maintainers": ["mmequignon"],
+    "license": "AGPL-3",
+    "installable": True,
+    "auto_install": True,
```
`"auto_install": False` no?

simahawk replied:
yes pls

sebalix commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,877 @@
+    _inherit = "base.shopfloor.process"
+    _name = "shopfloor.reception"
+    _usage = "reception"
+    _description = __doc__
```
A docstring explaining this scenario is welcome, as it is done in others.

simahawk replied:
A long one pls :)

mmequignon replied:
@jbaudoux I'll need help for that.

jbaudoux replied:
Here is what you can add in the docstring:

Process a picking and track progress by product.

Once a transfer is opened, you first need to select a product (you can scan its barcode or one of its packaging barcode). Then set the processed quantity. Put it in a internal PACK (this is optional but can be made mandatory by menu configuration), this PACK can be a new one (like an empty pallet) or an existing one you add products to (like a pallet you continue to fill in). And finally, the location where you put the product (iow. the location where is the transport trolley or pallet), unless you fill an existing PACK as its location was already defined when its first product was put on it.

In case of product tracked by lot, you will have to enter the lot number and its expiry date (unless it is already known by the system).

Moves are not validated as they are processed. It is the responsibility of the user to decide when to mark as done already processed lines.Any remaining lines will be pushed to a backorder.

sebalix commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,877 @@
+        return self._response_for_select_document(
+            message={
+                "message_type": "error",
+                "body": _("No picking found for the scanned packaging."),
```
As said, messages have to be moved to `actions/message.py`, or re-use if possible the existing messages in `shopfloor` module.
Also, I don't think the term `picking` should be used for user messages, we use operation or transfer (to check, I don't remember) in other places. Better to stick with Odoo naming when we share info with the user.

sebalix commented on shopfloor_reception/docs/reception_sequence_graph.mermaid:
```diff
@@ -0,0 +1,38 @@
+%%{init: {'theme': 'neutral' } }%%
```
Can you mention the name of the methods to call to switch from one state to another?
Like it is done in https://github.com/OCA/wms/blob/14.0/shopfloor/docs/delivery_diag_seq.png for instance

sebalix commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,877 @@
+                        }
+                    )
+                line.lot_id = lot.id
+                # TODO: Dunno why, I have to explicitely call the onchange method
```
Not sure to understand this comment?
It's not a good thing to call the onchange methods here anyway? Looking at the code of `product_expiry`, this onchange updates the `expiration_date` field of the line, that seems important.

mmequignon replied:
Yeah this is important. The thing is that changing the lot doesn't update the expiration date.
We have to force it, despite what suggests this piece of code https://github.com/odoo/odoo/blob/14.0/addons/product_expiry/models/stock_move_line.py#L29-L36

sebalix replied:
Onchange methods are a UI feature, they are never called when fields are updated on server-side, so it's normal to call them manually. This will be fixed on 16.0 with the new way to write "onchange" methods.

sebalix commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,877 @@
+          - set_lot: Error: expiration_date is required
+          - set_quantity: User clicked on the confirm button
+        """
+        picking = self.env["stock.picking"].browse(picking_id)
+        selected_lines = self.env["stock.move.line"].browse(selected_line_ids)
```
Do not hesitate to check if the ID of the records you are receiving really exist (with `exists()` method), like here: https://github.com/OCA/wms/blob/14.0/shopfloor/services/zone_picking.py#L768-L770

sebalix commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,877 @@
+                    if pack_location not in child_locations:
+                        # If the scanned package has a location that isn't a child
+                        # of the move dest, return an error
+                        message = {
+                            "message_type": "error",
+                            "body": _("You cannot place it here."),
+                        }
```
```suggestion
                        message = self.msg_store.dest_location_not_allowed()
```

sebalix commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,877 @@
+            if location:
+                if location not in child_locations:
+                    # Scanned location isn't a child of the move's dest location
+                    message = {
+                        "message_type": "error",
+                        "body": _("You cannot place it here."),
+                    }
```
```suggestion
                    message = self.msg_store.dest_location_not_allowed()
```

sebalix commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,877 @@
+                        "body": _("Put in this new PACK {}?").format(barcode),
+                    },
+                )
+            package = self.env["stock.quant.package"].create({"name": barcode})
```
The creation of a package wasn't optional (reading the specs)?

> Option on menu item:
> - Always put in Pack (SSCC)
>

mmequignon replied:
Well, regarding the specs, I should have put that in the shopfloor module.
I thought it would be smarter to create a new module. If the module is installed, then I suppose that this option is enabled by default.

sebalix replied:
It has its place in this module I think as we generally create packages when receiving the goods. And the option is enabled on the menu, so we could have two different menus to receive goods, with the option enabled only on one.
cc @jbaudoux

jbaudoux replied:
At this stage, it looks good, all the tests have been done and the pack can now be created

JuMiSanAr commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,877 @@
+
+    @property
+    def _schema_select_document(self):
+        return {"pickings": self.schemas._schema_list_of(self.schemas.picking())}
```
We need the lines in the response, so this needs to make use of _schema_stock_picking as discussed.

simahawk commented on shopfloor_reception/__manifest__.py:
```diff
@@ -0,0 +1,20 @@
+    "version": "14.0.1.0.0",
+    "category": "Inventory",
+    "website": "https://github.com/OCA/wms",
+    "author": "Camptocamp SA, Odoo Community Association (OCA)",
```
```suggestion
    "author": "Camptocamp, Odoo Community Association (OCA)",
```

simahawk commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,877 @@
+        return [("package_id.product_packaging_id", "=", packaging.id)]
+
+    def _domain_move_line_by_product(self, product):
+        return [("product_id", "=", product.id)]
```
must filter by state and by self.picking_types at least

simahawk commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,877 @@
+
+    def _data_for_stock_picking(self, picking):
+        data = self.data.picking(picking)
+        data.update({"move_lines": self.data.move_lines(picking.move_line_ids)})
```
```suggestion
        data.update({"move_lines": self._data_for_move_lines(picking.move_line_ids)})
```

simahawk commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,877 @@
+            message=message,
+        )
+
+    # ENDPOINTS
```
I find it weird to have all the endpoints a the bottom... but yeah...

simahawk commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,877 @@
+        if message:
+            return self._response_for_select_line(picking, message=message)
+
+        search = self._actions_for("search")
```
not a blocker but you could do it like this https://github.com/OCA/wms/blob/14.0/shopfloor/services/zone_picking.py#L594

simahawk commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,877 @@
+        )
+
+    def done_action(self, picking_id, confirmation=False):
+        picking = self.env["stock.picking"].browse(picking_id)
```
please use `_check_picking_status` when you lookup for pickings https://github.com/OCA/wms/blob/40f1e785a313f38f85b27871f9e81a3e2a4ae72a/shopfloor/services/service.py#L49

simahawk replied:
still not done :arrow_up:

simahawk commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,877 @@
+            return self._response_for_select_line(
+                picking, message=self.msg_store.need_confirmation()
+            )
+        backorders_before = picking.backorder_ids
```
maybe delegate to `_handle_backorders`?

mmequignon committed:
fixup! Add shopfloor_reception

simahawk commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,951 @@
+
+    # DATA METHODS
+
+    def _data_for_stock_picking(self, picking):
```
This seems sub optimal. We should apply the same parser to the whole recordset.
You should probably use `data_detail.pickings_detail()` which already contains data for the move lines

simahawk commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,951 @@
+    def _order_stock_picking(self):
+        return "scheduled_date ASC"
+
+    def _scan_document__scan_picking(self, barcode):
```
maybe replace "scan" w/ "by" in all these methods?
```suggestion
    def _scan_document__by_picking(self, barcode):
```

simahawk commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,951 @@
+        message = self._check_picking_status(picking)
+        if message:
+            return self._response_for_select_line(picking, message=message)
+        message = self._check_picking_status(picking)
+        if message:
+            return self._response_for_select_line(picking, message=message)
```
```suggestion
```

simahawk commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,951 @@
+          - set_quantity: User clicked on the confirm button
+        """
+        picking = self.env["stock.picking"].browse(picking_id)
+        selected_lines = self.env["stock.move.line"].browse(selected_line_ids)
+        message = self._check_picking_status(picking)
+        if message:
+            return self._response_for_set_lot(picking, selected_lines, message=message)
```
```suggestion
        message = self._check_picking_status(picking)
        if message:
            return self._response_for_set_lot(picking, selected_lines, message=message)
        selected_lines = self.env["stock.move.line"].browse(selected_line_ids).exists()
```

mmequignon replied:
can't do that, selected lines is sent in the response.

simahawk commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,951 @@
+        if lot_name:
+            for line in selected_lines:
+                lot = search.lot_from_scan(
+                    lot_name, products=selected_lines.mapped("product_id")
```
For each line you search for all the products of all lines.
I would collect all the line w/out a lot, collect all values for missing lots and create them all at once.
You can map line ids and lots in a dict.

JuMiSanAr replied:
We only deal with 1 line at a time now so this is no longer the case.

simahawk commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,951 @@
+        """
+        picking = self.env["stock.picking"].browse(picking_id)
+        selected_lines = self.env["stock.move.line"].browse(selected_line_ids)
+        message = self._check_picking_status(picking)
```
always check the status before doing anything else

mmequignon replied:
In all cases, I return a response with the selected lines, whether there's a message or not.

simahawk commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,951 @@
+        }
+
+    def _schema_stock_picking(self, lines_with_packaging=False):
+        schema = self.schemas.picking()
```
use picking_detail schema

simahawk replied:
?

JuMiSanAr replied:
It was updated to _schema_stock_picking_with_lines

simahawk replied:
you got my point ;)

simahawk commented on shopfloor_reception/tests/common.py:
```diff
@@ -0,0 +1,83 @@
+            line.package_id = package
+
+    @classmethod
+    def setUpClassVars(cls, *args, **kwargs):
```
please keep setup methods on top

simahawk replied:
???

jbaudoux commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,877 @@
+                return self._response_for_set_destination(picking, selected_lines)
+            location = search.location_from_scan(barcode)
+            if location:
+                if location not in child_locations:
```
I think its lacking a test if the pack is mandatory and already set

JuMiSanAr committed:
shopfloor_mobile_base: implement clearable input type in searchbar

mmequignon committed:
shopfloor: Add stock_picking_completion as dependency

In some cases in the shopfloor app, we need to display progression of
pickings or move lines. Instead of sending the whole move lines and
compute it in the frontend, send directly the completion.

mmequignon committed:
shopfloor_reception: Fix perf issues on start screen

mmequignon committed:
shopfloor_reception_mobile: Fix perf issues on start screen

JuMiSanAr committed:
fixup! shopfloor: Add stock_picking_completion as dependency

JuMiSanAr committed:
shopfloor_reception: add confirmation screen after mark as done

JuMiSanAr committed:
shopfloor: allow source_document as identifier for the search

JuMiSanAr committed:
shopfloor_reception: select_dest_package

JuMiSanAr committed:
shopfloor_reception: allow scanning source document

JuMiSanAr committed:
fixup! BSRAU-57: shopfloor_reception_mobile frontend

JuMiSanAr committed:
shopfloor_reception: update quantity when clicking new pack button

jbaudoux commented on shopfloor/__manifest__.py:
```diff
@@ -27,6 +27,7 @@
         #  OCA / stock-logistics-workflow
         "stock_quant_package_dimension",
         "stock_quant_package_product_packaging",
+        "stock_picking_completion",
```
Where is this module?
Rename to stock_picking_progress

mmequignon replied:
done.

JuMiSanAr committed:
fixup! shopfloor_reception: update quantity when clicking new pack button

mmequignon committed:
fixup! fixup! Add shopfloor_reception

mmequignon committed:
fixup! fixup! shopfloor: Add stock_picking_completion as dependency

mmequignon commented on shopfloor_reception/services/reception.py:
```diff
@@ -72,6 +72,11 @@ def _response_for_select_line(self, picking, message=None):
+        picking.user_id = self.env.user
+        data = {"picking": self._data_for_stock_picking(picking, with_lines=True)}
+        return self._response(next_state="confirm_done", data=data, message=message)
+
```
@JuMiSanAr could you add tests for that?

mmequignon commented on shopfloor/actions/search.py:
```diff
@@ -28,11 +30,19 @@ def package_from_scan(self, barcode):
+                ("origin", "=", barcode),
+                ("state", "=", "assigned"),
+            ]
+            domain = expression.OR([domain, source_document_domain])
```
Needs to be tested too

mmequignon commented on shopfloor_reception/services/reception.py:
```diff
@@ -319,10 +319,27 @@ def _response_for_set_destination(self, picking, lines, message=None):
+            packages.with_context(picking_id=picking.id).sorted(),
+            picking=picking,
+            with_packaging=True,
+        )
```
we need tests here too

mmequignon commented on shopfloor_reception/services/reception.py:
```diff
@@ -158,8 +158,15 @@ def _order_stock_picking(self):
     def _scan_document__scan_picking(self, barcode):
         search = self._actions_for("search")
-        picking = search.picking_from_scan(barcode)
+        picking = search.picking_from_scan(barcode, True)
```
Since this is a named argument, add the key arg, it will make this easier to read
```python
        picking = search.picking_from_scan(barcode, accept_source_document=True)
```
Anyway, why making this argument optional? Can't see the point

sebalix replied:
That's a good thing to make it optional by default, we already had done this on product scan in the past to also accept product packaging barcode, and with time we had to refactor it to make it optional.
Now we can debate: should the default value on the method be `True` or `False`...

JuMiSanAr replied:
Since at this stage the only scenario that accepts SO is reception (last to arrive), I set it to false by default in all other cases. Not sure if all other scenarios will have to accept it as well in the future? @jbaudoux

JuMiSanAr committed:
fixup! shopfloor_reception: allow scanning source document

JuMiSanAr committed:
shopfloor_reception: redirect to select_line after select_dest_package

sebalix commented on shopfloor/actions/search.py:
```diff
@@ -28,11 +30,20 @@ def package_from_scan(self, barcode):
         return model.search([("name", "=", barcode)], limit=1)
 
-    def picking_from_scan(self, barcode):
+    def picking_from_scan(self, barcode, accept_source_document=False):
```
We have the same kind of parameter named `use_packaging` when scanning a product, so I would name it the same, and based on the technical field name, not its label:
```suggestion
    def picking_from_scan(self, barcode, use_origin=False):
```

JuMiSanAr committed:
shopfloor_reception: refactor & general improvements frontend

JuMiSanAr committed:
shopfloor_reception: ask for confirmation to create new package

JuMiSanAr committed:
fixup! fixup! shopfloor: Add stock_picking_completion as dependency

JuMiSanAr committed:
shopfloor_reception frontend: use progress to get summary picking class color

JuMiSanAr committed:
shopfloor_reception: refactor to selected_move_line

The previous version always assumed that there could be multiple selected lines being handled.
This is never the case in most of the screens, so now only one is expected

JuMiSanAr committed:
shopfloor_reception backend: fixes for datas & schemas

JuMiSanAr committed:
shopfloor_reception frontend: fixes to frontend UI

JuMiSanAr committed:
shopfloor_reception: autofill expiry date when a lot is selected in set_lot screen

JuMiSanAr committed:
fixup! shopfloor_reception: refactor to selected_move_line

JuMiSanAr committed:
fixup! shopfloor_reception frontend: fixes to frontend UI

JuMiSanAr committed:
fixup! shopfloor_reception: autofill expiry date when a lot is selected in set_lot screen

JuMiSanAr committed:
shopfloor_reception: fix bug with backend not sending back data

JuMiSanAr committed:
fixup! fixup! shopfloor_reception frontend: fixes to frontend UI

JuMiSanAr committed:
shopfloor_reception frontend: sort lines to select by progress

JuMiSanAr committed:
shopfloor_reception: fix unit tests

JuMiSanAr committed:
shopfloor_reception: introduce new unit tests

JuMiSanAr committed:
shopfloor: update tests after introducing shopfloor_reception scenario

simahawk commented on shopfloor/actions/data.py:
```diff
@@ -49,6 +49,7 @@ def _picking_parser(self):
             "bulk_line_count",
             "total_weight:weight",
             "scheduled_date",
+            "progress",
```
Are we sure we need to load this information for ALL the pickings and for all the cases?
AFAIS on https://github.com/OCA/stock-logistics-workflow/pull/1116 it can be heavy (unless you improve perf)

JuMiSanAr replied:
For the moment the performance is good. 

Should I create a different parser just for this?

simahawk replied:
as we said, pls make it optional

simahawk commented on shopfloor/actions/data.py:
```diff
@@ -64,8 +65,10 @@ def package(self, record, picking=None, with_packaging=False, **kw):
-            lines = picking.move_line_ids.filtered(lambda l: l.package_id == record)
+            lines = picking.move_line_ids.filtered(
+                lambda l: l.result_package_id == record
+                and l.state in ["partially_available", "assigned", "done"]
```
@JuMiSanAr this is the change the required the tests update, not the reception scenario introduction

JuMiSanAr replied:
Not sure I understand what you mean here.

simahawk replied:
I mean that the fix for the tests has been committed into the reception but it has nothing to deal with it

simahawk commented on shopfloor/actions/data.py:
```diff
@@ -131,7 +134,7 @@ def lots(self, record, **kw):
     @property
     def _lot_parser(self):
-        return self._simple_record_parser() + ["ref"]
+        return self._simple_record_parser() + ["ref", "expiration_date"]
```
and this one

simahawk commented on shopfloor/actions/data.py:
```diff
@@ -176,6 +179,7 @@ def _move_line_parser(self):
                 "move_id:priority",
                 lambda rec, fname: rec.move_id.priority or "",
             ),
+            "progress",
```
and this one

simahawk commented on shopfloor_mobile/static/wms/src/components/scenario_picking_detail/picking_summary.js:
```diff
@@ -114,53 +116,87 @@ Vue.component("picking-summary-content", {
                     </span>
-                    {{ record.title }}
+                    <strong>
+                        {{ options.group_header_title_key ? pkg_type.records[0].product[options.group_header_title_key] : record.title }}
```
```suggestion
                        {{ get_group_title(rec, pkg_type) }}
```

simahawk commented on shopfloor_mobile/static/wms/src/components/scenario_picking_detail/picking_summary.js:
```diff
@@ -114,53 +116,87 @@ Vue.component("picking-summary-content", {
     template: `
         <div class="summary-content-item">
-            <v-list-item-title>
+            <v-list-item-title v-if="options.show_title">
```
this should default to true and is not the case now, no?

JuMiSanAr replied:
It does (in the List component).

simahawk commented on shopfloor_mobile/static/wms/src/wms_utils.js:
```diff
@@ -86,6 +86,47 @@ export class WMSUtils {
+        const self = this;
+        // {'key': 'no-group', 'title': '', 'records': []}
+        options = _.defaults(options || {}, {
+            group_key: "product",
```
what's the point of making the group key configurable?

simahawk replied:
Do you need all these options?

simahawk commented on shopfloor_mobile_base/static/wms/src/components/datepicker/date_picker.js:
```diff
@@ -0,0 +1,48 @@
+    },
+    watch: {
+        date: function () {
+            this.$root.trigger("date_picker_selected", this.date);
```
Either you use `event_hub.$emit` or `this.$root.trigger`.
I don't think you need an event which has a global state here -> you don't need `$root.trigger`.

simahawk commented on shopfloor_mobile_base/static/wms/src/components/datepicker/date_picker.js:
```diff
@@ -0,0 +1,48 @@
+        },
+    },
+    mounted() {
+        event_hub.$on("datepicker:lotselected", (lot) => {
```
:fire: 

This is a specific feature that should be implemented by the consumer of this widget.
You should have something like

`<date-picker-input v-on:date_selected="on_date_selected"...`

simahawk replied:
Also, this change should be move to its own commit and separated

simahawk commented on shopfloor_mobile_base/static/wms/src/components/searchbar/searchbar.js:
```diff
@@ -52,6 +52,11 @@ export var Searchbar = Vue.component("searchbar", {
+        // uses the clearable input type logic (using an X icon to clear the input)
+        clearable_input_type: {
+            type: Boolean,
+            default: false,
```
can be true likely?

JuMiSanAr replied:
This is a very specific thing of this filter searchbar (as opposed to the standard scan searchbar we use everywhere), I wouldn't set it as true by default as it can mess with the rest of the searchbars.

I've given it a try and it alters the behaviour of the rest of the searchbars in the app.

simahawk commented on shopfloor_reception/data/shopfloor_scenario_data.xml:
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<!-- Copyright 2022 Camptocamp SA
+     License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl). -->
+<odoo>
```
```suggestion
<odoo noupdate="1">
```

simahawk commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,1079 @@
+
+    Process a picking and track progress by product.
+
+    Once a transfer is opened, you first need to select a product (you can scan its
```
can you rephrase this and use an ordered list?

simahawk commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,1079 @@
+    def _response_for_confirm_new_package(
+        self, picking, line, new_package_name, message=None
+    ):
+        picking.user_id = self.env.user
```
I would not set the user here. This should be one explicitly by the endpoint, not by the method collecting data for the response.

simahawk commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,1079 @@
+        move_lines = self._move_line_by_packaging(packaging)
+        pickings = move_lines.mapped("move_id.picking_id")
+        if len(pickings) == 1:
+            pickings.user_id = self.env.user
```
```suggestion
            self._assign_user_to_picking(picking)
```
you know where to use it ;)

simahawk commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,1079 @@
+                          single correspondance. Not tracked product
+        """
+        self._actions_for("search")
+        handlers = [
+            self._scan_document__by_picking,
+            self._scan_document__by_product,
+            self._scan_document__by_packaging,
+        ]
```
```suggestion
        handlers = (
            self._scan_document__by_picking,
            self._scan_document__by_product,
            self._scan_document__by_packaging,
        )
```
always use tuples when you don't expect modifications

simahawk commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,1079 @@
+        message = self._check_picking_status(picking)
+        if message:
+            return self._response_for_select_line(picking, message=message)
+        handlers = [
```
same

simahawk commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,1079 @@
+
+    def _use_handlers(self, handlers, *args, **kwargs):
+        for handler in handlers:
+            response = handler(*args, **kwargs)
+            if response:
+                return response
```
```suggestion
            return handler(*args, **kwargs)
```

JuMiSanAr replied:
Hmmm wouldn't this always return whatever the first handler returns, and stop the loop there?

simahawk commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,1079 @@
+            product = selected_line.product_id
+            lot = search.lot_from_scan(lot_name, products=product)
+            if not lot:
+                lot = self.env["stock.production.lot"].create(
+                    {
+                        "name": lot_name,
+                        "product_id": product.id,
+                        "company_id": self.env.company.id,
+                        "use_expiration_date": product.use_expiration_date,
+                    }
+                )
```
```suggestion
                lot = self.env["stock.production.lot"].create(
                    self._create_lot_values(product, lot_name)
                )
```

simahawk commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,1079 @@
+                    }
+                )
+            selected_line.lot_id = lot.id
+            selected_line._onchange_lot_id()
```
use onchange_helper instead?

simahawk commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,1079 @@
+        return self._response_for_set_lot(picking, selected_line)
+
+    def set_lot_confirm_action(self, picking_id, selected_line_id):
+        picking = self.env["stock.picking"].browse(picking_id)
```
I repeat myself: every time you lookup for a picking, check its status.

simahawk commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,1079 @@
+    def set_lot_confirm_action(self, picking_id, selected_line_id):
+        picking = self.env["stock.picking"].browse(picking_id)
+        selected_line = self.env["stock.move.line"].browse(selected_line_id)
+        message = self._check_expiry_date(selected_line)
```
this is a check, is not meant to give you a msg IMO

JuMiSanAr replied:
it's the same as _check_picking_status isn't it?

simahawk commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,1079 @@
+            selected_line.qty_done = quantity
+            return self._response_for_set_quantity(picking, selected_line)
+        if barcode:
+            handlers = [
+                self._set_quantity__by_product,
+                self._set_quantity__by_packaging,
+                self._set_quantity__by_package,
+                self._set_quantity__by_location,
+            ]
```
```suggestion
            handlers = (
                self._set_quantity__by_product,
                self._set_quantity__by_packaging,
                self._set_quantity__by_package,
                self._set_quantity__by_location,
            )
```

simahawk commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,1079 @@
+            selected_line.location_dest_id = location
+            return self._response_for_select_line(picking)
+        elif location in pick_type_child_locations:
+            # If location is a child of picking types's dest lcoation,
```
```suggestion
            # If location is a child of picking types's dest location,
```

simahawk commented on shopfloor_reception_mobile/static/src/scenario/reception.js:
```diff
@@ -0,0 +1,723 @@
+            );
+        },
+        on_select_document: function (selected) {
+            this.$root.trigger("scan", selected.name);
```
why do you trigger an event from another event?

simahawk commented on shopfloor_reception_mobile/static/src/scenario/reception.js:
```diff
@@ -0,0 +1,723 @@
+        on_select_document: function (selected) {
+            this.$root.trigger("scan", selected.name);
+        },
+        on_mark_done: function () {
```
same

simahawk commented on shopfloor_reception_mobile/static/src/scenario/reception.js:
```diff
@@ -0,0 +1,723 @@
+            const next_unhandled_line = this.state.data.selected_move_lines.find(
+                (line) => line.qty_done < line.quantity
+            );
+            return this._get_selected_line_ids([next_unhandled_line]);
```
`_get_selected_line_ids` is used only here and only w/ one record, why do you need a list and loop on it to get the id?

simahawk commented on shopfloor_reception_mobile/static/src/scenario/reception.js:
```diff
@@ -0,0 +1,723 @@
+                        );
+                    },
+                    on_cancel: () => {
+                        // TODO: endpoint missing in backend
```
??

JuMiSanAr replied:
It was decided with @jbaudoux that we'd leave this implementation in the backend for a second version of the scenario.
I left this in the frontend so that it's easier to implement later.

simahawk commented on shopfloor_reception_mobile/static/src/scenario/reception.js:
```diff
@@ -0,0 +1,723 @@
+                        );
+                    },
+                    on_cancel: () => {
+                        // TODO: endpoint missing in backend
```
???

simahawk commented on shopfloor_reception_mobile/static/src/scenario/reception.js:
```diff
@@ -0,0 +1,723 @@
+
+export default Reception;
+
+// TODO: handle all confirmations in the scenario
```
???

sebalix commented on shopfloor/actions/search.py:
```diff
@@ -28,11 +30,20 @@ def package_from_scan(self, barcode):
+        source_document_domain = [
+            "&",
+            ("origin", "=", barcode),
+            ("state", "=", "assigned"),
```
Why are we checking the state when searching against `origin` and not with `name`? IMO we should not check the `state` of the transfer at all in this method, it's the job of the `_check_picking_status` general method. So we could write:
```python
    def picking_from_scan(self, barcode, use_origin=False):
        model = self.env["stock.picking"]
        if not barcode:
            return model.browse()
        domain = [("name", "=", barcode)]
        if use_origin:
            origin_domain = [("origin", "=", barcode)]
            domain = expression.OR([domain, origin_domain])
        return model.search(domain)
```

sebalix commented on shopfloor_reception/docs/reception_sequence_graph.mermaid:
```diff
@@ -0,0 +1,61 @@
+	select_dest_package ->> select_line: User scanned a valid package
+	select_dest_package ->> select_dest_package: Warning: User scanned an unknown barcode. Confirm to create one.
+	select_dest_package ->> select_dest_package: Error: User scanned a non-empty package
+    end
```
IMO I find the graph generated with mermaid not readable at all compared to what we were using in other scenarios (PlantUML)... need to zoom in and scroll a lot (and the image quality is not so good).
But don't spend time on this.

sebalix commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,1079 @@
+    Keep [the sequence diagram](../docs/reception_sequence_graph.mermaid)
+    up-to-date if you change endpoints.
+
+    Process a picking and track progress by product.
```
Better to fit with what we are processing with this scenario, and re-use Odoo terms:
```suggestion
    Process a receipt transfer and track progress by product.
```

JuMiSanAr replied:
Updated.
I'm assuming this scenario description is the only place we should update, as this is referred to as "picking" in the rest of the module right?

sebalix replied:
Yes, only in the docstring.
I think the right fix would be to rename `stock.picking` to `stock.transfer`, as picking is a type of transfer, but we won't win this battle I guess :)

sebalix commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,1079 @@
+        # selected_line
+        # picking
+        search = self._actions_for("search")
+        product = search.product_from_scan(barcode, use_packaging=False)
```
NOTE for ourself: remove `use_packaging=False` once https://github.com/OCA/wms/pull/483/commits/f3cda915e9659a3cbadc66ee4c8781f76bb22281 is merged (#483)

sebalix commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,1079 @@
+        # selected_line
+        # picking
+        search = self._actions_for("search")
+        product = search.product_from_scan(barcode, use_packaging=False)
+        if product:
+            selected_line.qty_done += 1
```
We should check if the product scanned matches with the line, no? Right now I could scan product A while the one on the selected line is B.

sebalix commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,1079 @@
+            return self._select_line_from_packaging(picking, packaging)
+
+    def _set_quantity__by_product(self, picking, selected_line, barcode):
+        # needed values:
+        # selected_line
+        # picking
```
```suggestion
```

sebalix commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,1079 @@
+    def _set_quantity__by_packaging(self, picking, selected_line, barcode):
+        search = self._actions_for("search")
+        packaging = search.packaging_from_scan(barcode)
+        if packaging:
+            selected_line.qty_done += packaging.qty
```
Same remark than :arrow_up: check that the packaging product is the same than the line.

sebalix commented on shopfloor_reception/services/reception.py:
```diff
@@ -0,0 +1,1079 @@
+        return self._response(next_state="select_document", data=data, message=message)
+
+    def _response_for_set_lot(self, picking, line, message=None):
+        # If the product isn't tracked, move to the next screen
```
```suggestion
```
As that's not what this method is doing.

JuMiSanAr committed:
shopfloor_reception: add secondary sorting of the pickings

The pickings were being sorted by scheduled_date only, which means that
if there are two pickings with the exact same scheduled_date,
we had no control on the order.
Now they are sorted by id in that case.

JuMiSanAr committed:
shopfloor_reception frontend: display group-shared information in summary-picking header

JuMiSanAr committed:
shopfloor_reception frontend: improve date-picker-input

JuMiSanAr committed:
fixup! fixup! fixup! Add shopfloor_reception

JuMiSanAr committed:
shopfloor_reception: check picking status everywhere

JuMiSanAr committed:
shopfloor_reception: update warning message

JuMiSanAr committed:
shopfloor_reception: use method to assign user

JuMiSanAr committed:
fixup! shopfloor_reception: add secondary sorting of the pickings

JuMiSanAr committed:
shopfloor_reception: update comments and minor refactors

JuMiSanAr committed:
shopfloor_reception frontend: cleanup

JuMiSanAr committed:
fixup! shopfloor_reception frontend: cleanup

JuMiSanAr committed:
shopfloor_reception: fix incorrect qty after set_quantiy scan

After scanning anything on the set_quantity screen,
the resulting qty_done was set to be equal to the qty of the picker.
Now, it's the qty of the picker + 1 if scanning a product
or + packaging qty if scanning a packaging.

JuMiSanAr committed:
shopfloor: make progress data key optional for picking

JuMiSanAr committed:
shopfloor_reception: check correct product id on scan

JuMiSanAr committed:
shopfloor_reception: update comments & roadmap

JuMiSanAr committed:
fixup! shopfloor_reception: fix unit tests

sebalix commented:
Closing in favor of #500
